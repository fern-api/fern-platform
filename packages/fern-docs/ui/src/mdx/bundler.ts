import type * as FernDocs from "@fern-api/fdr-sdk/docs";
import type { FernSerializeMdxOptions } from "./types";

export type MDX_SERIALIZER = {
  (
    content: string,
    options?: FernSerializeMdxOptions
  ): Promise<FernDocs.MarkdownText>;
  (
    content: string | undefined,
    options?: FernSerializeMdxOptions
  ): Promise<FernDocs.MarkdownText | undefined>;
};
