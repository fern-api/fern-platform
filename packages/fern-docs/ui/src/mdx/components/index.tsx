import { RemoteFontAwesomeIcon } from "@fern-docs/components";
import type { MDXComponents } from "mdx/types";
import dynamic from "next/dynamic";
import { ComponentProps, PropsWithChildren, ReactElement } from "react";
import {
  FernErrorBoundary,
  FernErrorBoundaryProps,
  FernErrorTag,
} from "../../components/FernErrorBoundary";
import { SidebarSearchBar } from "../../sidebar/SidebarSearchBar";
import { Accordion, AccordionGroup } from "./accordion";
import { Availability } from "./availability";
import { Badge } from "./badge";
import { Bleed } from "./bleed";
import { Button, ButtonGroup } from "./button";
import {
  Callout,
  CheckCallout,
  ErrorCallout,
  InfoCallout,
  LaunchNoteCallout,
  NoteCallout,
  SuccessCallout,
  TipCallout,
  WarningCallout,
} from "./callout";
import { Card, CardGroup } from "./card";
import { ClientLibraries } from "./client-libraries";
import { CodeBlock } from "./code/CodeBlock";
import { CodeGroup } from "./code/CodeGroup";
import { Column, ColumnGroup } from "./columns";
import { Feature } from "./feature";
import { Frame } from "./frame";
import { A, Embed, HeadingRenderer, Image, Li, Ol, Strong, Ul } from "./html";
import { Table } from "./html-table";
import { If } from "./if";
import { IFrame } from "./iframe";
import { Mermaid } from "./mermaid";
import { ParamField } from "./mintlify";
import { ReferenceLayoutAside, ReferenceLayoutMain } from "./reference-layout";
import { EndpointRequestSnippet, EndpointResponseSnippet } from "./snippets";
import { Step, StepGroup } from "./steps";
import { Tab, TabGroup } from "./tabs";
import { Tooltip } from "./tooltip";

const ElevenLabsWaveform = dynamic(() =>
  import("./waveform/WaveformComplex").then((mod) => mod.default)
);

const FERN_COMPONENTS = {
  Accordion,
  AccordionGroup,
  Availability,
  Badge,
  Bleed,
  Button,
  ButtonGroup,
  Callout,
  Card,
  CardGroup,
  ClientLibraries,
  CodeBlock,
  CodeGroup, // note: alias is handled in rehypeFernCode
  Column,
  ColumnGroup,
  EndpointRequestSnippet,
  EndpointResponseSnippet,
  Feature,
  Frame,
  Icon: RemoteFontAwesomeIcon,
  If,
  Mermaid,
  ParamField,
  SearchBar: SidebarSearchBar,
  Step,
  StepGroup,
  Tab,
  TabGroup,
  Tooltip,

  // callout aliases
  Info: InfoCallout,
  Warning: WarningCallout,
  Success: SuccessCallout,
  Error: ErrorCallout,
  Note: NoteCallout,
  Tip: TipCallout,
  Check: CheckCallout,
  LaunchNote: LaunchNoteCallout,

  // deprecated, aliased for backwards compatibility
  Accordions: AccordionGroup,
  Cards: CardGroup,
  CodeBlocks: CodeGroup,
  ElevenLabsWaveform,
  Expandable: Accordion,
  Tabs: TabGroup,
};

// internal-use only
const INTERNAL_COMPONENTS = {
  ReferenceLayoutMain,
  ReferenceLayoutAside,

  // error boundary
  FernErrorBoundary: (
    props: PropsWithChildren<Pick<FernErrorBoundaryProps, "error" | "fallback">>
  ): ReactElement => <FernErrorBoundary {...props} />,
};

const HTML_COMPONENTS: MDXComponents = {
  a: A,
  embed: Embed,
  h1: (props) => HeadingRenderer(1, props),
  h2: (props) => HeadingRenderer(2, props),
  h3: (props) => HeadingRenderer(3, props),
  h4: (props) => HeadingRenderer(4, props),
  h5: (props) => HeadingRenderer(5, props),
  h6: (props) => HeadingRenderer(6, props),
  img: Image,
  li: Li,
  ol: Ol,
  strong: Strong,
  table: Table,
  ul: Ul,
};

const ALIASED_HTML_COMPONENTS = {
  A,
  Embed,
  H1: (props: ComponentProps<"h1">): ReactElement => HeadingRenderer(1, props),
  H2: (props: ComponentProps<"h2">): ReactElement => HeadingRenderer(2, props),
  H3: (props: ComponentProps<"h3">): ReactElement => HeadingRenderer(3, props),
  H4: (props: ComponentProps<"h4">): ReactElement => HeadingRenderer(4, props),
  H5: (props: ComponentProps<"h5">): ReactElement => HeadingRenderer(5, props),
  H6: (props: ComponentProps<"h6">): ReactElement => HeadingRenderer(6, props),
  IFrame,
  Image,
  Li,
  Ol,
  Strong,
  Table,
  Ul,
};

export const MDX_COMPONENTS: MDXComponents = {
  ...FERN_COMPONENTS,
  ...INTERNAL_COMPONENTS,
  ...HTML_COMPONENTS,
  ...ALIASED_HTML_COMPONENTS,
};

export function createMdxComponents(jsxElements: string[]): MDXComponents {
  return {
    // spread in jsx elements that may be unsupported
    ...jsxElements.reduce<Record<string, () => ReactElement>>(
      (acc, jsxElement) => {
        acc[jsxElement] = () => (
          <FernErrorTag
            component={jsxElement}
            error={`Unsupported JSX tag: <${jsxElement} />`}
          />
        );
        return acc;
      },
      {}
    ),
    // then, spread in the supported components
    ...MDX_COMPONENTS,
  };
}
