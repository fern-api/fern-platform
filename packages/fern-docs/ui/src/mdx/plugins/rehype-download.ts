import {
  CONTINUE,
  Hast,
  isMdxJsxAttribute,
  isMdxJsxElementHast,
  SKIP,
  visit,
} from "@fern-docs/mdx";

/**
 * The goal of this plugin is to squeeze the <Download> element into a <Button> element with the `href` and `download` attributes set.
 * This is to avoid <a><button /></a> which is invalid HTML.
 */
export function rehypeDownload() {
  return (ast: Hast.Root) => {
    visit(ast, (node, index, parent) => {
      if (!isMdxJsxElementHast(node) || !parent || index == null) {
        return CONTINUE;
      }

      if (
        node.name === "Download" &&
        node.children.length === 1 &&
        node.children[0]
      ) {
        const child = node.children[0];
        if (isMdxJsxElementHast(child) && child.name === "Button") {
          const srcAttr = node.attributes
            .filter(isMdxJsxAttribute)
            .find((attr) => attr.name === "src");

          if (!srcAttr) {
            return CONTINUE;
          }

          const filenameAttr = node.attributes
            .filter(isMdxJsxAttribute)
            .find((attr) => attr.name === "filename");

          // inject the src attribute as a href
          child.attributes.push({
            type: "mdxJsxAttribute",
            name: "href",
            value: srcAttr.value,
          });

          // force the button to be a download
          child.attributes.push({
            type: "mdxJsxAttribute",
            name: "download",
            value: filenameAttr?.value,
          });

          parent.children[index] = child;

          return SKIP;
        }
      }
      return CONTINUE;
    });
  };
}
