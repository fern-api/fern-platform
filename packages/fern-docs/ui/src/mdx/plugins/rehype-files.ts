import type { Hast } from "@fern-docs/mdx";
import {
  isMdxJsxAttribute,
  isMdxJsxElementHast,
  mdxJsxAttributeToString,
  visit,
} from "@fern-docs/mdx";
import type { ImageData } from "../../atoms/types";

export interface RehypeFilesOptions {
  replaceSrc?(src: string): ImageData | undefined;
}

/**
 * This plugin will replace the src attribute with a new value, and add height and width attributes if they are provided.
 *
 * This is designed to handle injecting filesV2 urls into the following elements:
 * - img
 * - video
 * - audio
 * - source
 * - track
 * - embed
 * - script
 *
 * @param options - the options for the plugin
 * @returns a function that will transform the tree
 */
export function rehypeFiles(
  options: RehypeFilesOptions
): (tree: Hast.Root) => void {
  return function (tree: Hast.Root): void {
    visit(tree, (node) => {
      if (isMdxJsxElementHast(node)) {
        const attributes = node.attributes.filter(isMdxJsxAttribute);

        // TODO: do we need to add support for `href` and `<object data=...>`?
        const srcAttribute = attributes.find((attr) => attr.name === "src");

        if (srcAttribute == null) {
          return;
        }

        const src = mdxJsxAttributeToString(srcAttribute);
        if (!src) {
          console.warn(
            `[rehype-files]: src attribute is not parseable for ${node.name}`
          );
          return;
        }
        const {
          src: newSrc,
          height,
          width,
          blurDataURL,
        } = options.replaceSrc?.(src) ?? {};

        if (newSrc != null) {
          srcAttribute.value = newSrc;
        }

        if (
          height != null &&
          !attributes.find((attr) => attr.name === "height")
        ) {
          node.attributes.unshift({
            name: "height",
            value: String(height),
            type: "mdxJsxAttribute",
          });
        }

        if (
          width != null &&
          !attributes.find((attr) => attr.name === "width")
        ) {
          node.attributes.unshift({
            name: "width",
            value: String(width),
            type: "mdxJsxAttribute",
          });
        }

        if (
          blurDataURL &&
          node.name?.toLowerCase().startsWith("im") &&
          !attributes.find((attr) => attr.name === "blurDataURL")
        ) {
          node.attributes.unshift({
            name: "blurDataURL",
            value: blurDataURL,
            type: "mdxJsxAttribute",
          });
        }
      } else if (node.type === "element") {
        // TODO: do we need to add support for `href` and `<object data=...>`?
        const srcAttribute = node.properties?.src;

        if (typeof srcAttribute !== "string") {
          return;
        }

        // do something to the src attribute
        const {
          src: newSrc,
          height,
          width,
          // blurDataURL, should this be handled here?
        } = options.replaceSrc?.(srcAttribute) ?? {};

        if (newSrc != null) {
          node.properties.src = newSrc;
        }

        if (height != null && !node.properties.height) {
          node.properties.height = height;
        }

        if (width != null && !node.properties.width) {
          node.properties.width = width;
        }
      }
    });
  };
}
