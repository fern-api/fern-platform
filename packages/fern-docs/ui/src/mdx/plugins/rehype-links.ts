import {
  isMdxJsxAttribute,
  isMdxJsxElementHast,
  mdxJsxAttributeToString,
  visit,
  type Hast,
} from "@fern-docs/mdx";

export interface RehypeLinksOptions {
  replaceHref?: (href: string) => string | undefined;
}

export function rehypeLinks({ replaceHref }: RehypeLinksOptions = {}): (
  ast: Hast.Root
) => void {
  return (ast) => {
    if (replaceHref == null) {
      return;
    }
    visit(ast, (node) => {
      if (node.type === "element" && node.tagName === "a") {
        const href = node.properties.href;
        if (typeof href !== "string") {
          return;
        }
        const newHref = replaceHref(href);
        if (newHref == null) {
          return;
        }
        node.properties.href = newHref;
      }

      // TODO handle nested attributes and mdx expressions
      if (isMdxJsxElementHast(node)) {
        const attributes = node.attributes.filter(isMdxJsxAttribute);
        const hrefAttribute = attributes.find((attr) => attr.name === "href");
        if (hrefAttribute == null) {
          return;
        }
        const href = mdxJsxAttributeToString(hrefAttribute);
        if (!href) {
          return;
        }
        const newHref = replaceHref(href);
        if (newHref == null) {
          return;
        }
        hrefAttribute.value = newHref;
      }
    });
  };
}
