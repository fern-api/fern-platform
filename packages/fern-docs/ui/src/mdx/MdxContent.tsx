import type * as FernDocs from "@fern-api/fdr-sdk/docs";
import dynamic from "next/dynamic";
import React, { memo } from "react";
import { FernErrorBoundary } from "../components/FernErrorBoundary";
import { FrontmatterContextProvider } from "../contexts/frontmatter";

export declare namespace MdxContent {
  export interface Props {
    mdx: FernDocs.MarkdownText | FernDocs.MarkdownText[] | undefined;
    fallback?: React.ReactNode;
  }
}

const MdxBundlerComponent = dynamic(
  () =>
    import("./bundlers/mdx-bundler-component").then(
      (mod) => mod.MdxBundlerComponent
    ),
  { ssr: true }
);

const NextMdxRemoteComponent = dynamic(
  () =>
    import("./bundlers/next-mdx-remote-component").then(
      (mod) => mod.NextMdxRemoteComponent
    ),
  { ssr: true }
);

function isMdxEmpty(
  mdx: FernDocs.MarkdownText | FernDocs.MarkdownText[] | undefined
): boolean {
  if (!mdx) {
    return true;
  }

  if (typeof mdx === "string") {
    return mdx.trim().length === 0;
  }

  if (Array.isArray(mdx)) {
    return mdx.length === 0 || mdx.every(isMdxEmpty);
  }

  return mdx.code.trim().length === 0;
}

export const MdxContent = memo<MdxContent.Props>(function MdxContent({
  mdx,
  fallback,
}) {
  if (isMdxEmpty(mdx) || mdx == null) {
    return fallback;
  }

  if (typeof mdx === "string") {
    return mdx;
  }

  if (Array.isArray(mdx)) {
    return (
      <>
        {mdx.map((mdx, index) => (
          <MdxContent key={index} mdx={mdx} />
        ))}
      </>
    );
  }

  const MdxComponent =
    mdx.engine === "mdx-bundler" ? MdxBundlerComponent : NextMdxRemoteComponent;

  return (
    <FernErrorBoundary component="MdxContent">
      <FrontmatterContextProvider value={mdx.frontmatter}>
        <MdxComponent {...mdx} />
      </FrontmatterContextProvider>
    </FernErrorBoundary>
  );
});
