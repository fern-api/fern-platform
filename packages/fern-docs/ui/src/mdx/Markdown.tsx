import type * as FernDocs from "@fern-api/fdr-sdk/docs";
import clsx from "clsx";
import { memo, type ReactNode } from "react";
import { isMdxEmpty, MdxContent } from "./MdxContent";

export declare namespace Markdown {
  export interface Props {
    title?: ReactNode;

    // mdx: FernDocs.MarkdownText | FernDocs.MarkdownText[] | undefined;
    mdx: FernDocs.MarkdownText | undefined;
    className?: string;
    size?: "xs" | "sm" | "lg";

    /**
     * Fallback content to render if the MDX is empty
     */
    fallback?: ReactNode;

    fallbackAsChild?: boolean;
  }
}

export const Markdown = memo<Markdown.Props>(
  ({ title, mdx, className, size, fallback, fallbackAsChild }) => {
    // If the MDX is empty, return null
    if (!fallback && isMdxEmpty(mdx)) {
      return false;
    }

    if (fallbackAsChild) {
      return fallback;
    }

    return (
      <div
        className={clsx(
          className,
          "prose dark:prose-invert max-w-none break-words",
          {
            "whitespace-pre-wrap": typeof mdx === "string",
            "prose-base": size == null,
            "prose-sm dark:prose-invert-sm !text-xs": size === "xs",
            "prose-sm dark:prose-invert-sm": size === "sm",
            "prose-lg": size === "lg",
          }
        )}
      >
        {title}
        <MdxContent mdx={mdx} fallback={fallback} />
      </div>
    );
  }
);

Markdown.displayName = "Markdown";
