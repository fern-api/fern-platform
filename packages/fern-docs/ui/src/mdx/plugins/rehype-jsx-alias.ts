import { isMdxJsxElementHast } from "@fern-docs/mdx";
import type { Root } from "hast";
import { visit } from "unist-util-visit";

interface Options {
  aliases: Record<string, string>;
}

export function rehypeJsxAlias(
  options: Options = { aliases: {} }
): (tree: Root) => void {
  return function (tree: Root): void {
    // map intrinsic HTML components to their Fern equivalents
    visit(tree, (node) => {
      if (isMdxJsxElementHast(node) && node.name) {
        const alias = options.aliases[node.name];
        if (alias != null) {
          node.name = alias;
        }
      }
    });
  };
}
