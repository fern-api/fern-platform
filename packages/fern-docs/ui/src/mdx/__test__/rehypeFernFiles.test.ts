import { Hast, MdxJsxAttribute, isMdxJsxAttribute } from "@fern-docs/mdx";
import { setDimension } from "../plugins/rehype-files";

describe("setDimension", () => {
  it("should preserve image with both width and height set", () => {
    const node = createElement(100, 100);
    setDimension(node, node.attributes.filter(isMdxJsxAttribute), 200, 200);
    expect(node).toMatchSnapshot();
  });

  it("should properly assign width when only height is set", () => {
    const node = createElement(undefined, 100);
    setDimension(node, node.attributes.filter(isMdxJsxAttribute), 200, 200);
    expect(node).toMatchSnapshot();
  });

  it("should properly assign height when only width is set", () => {
    const node = createElement(100, undefined);
    setDimension(node, node.attributes.filter(isMdxJsxAttribute), 200, 200);
    expect(node).toMatchSnapshot();
  });

  it("should set width and height when neither is set", () => {
    const node = createElement();
    setDimension(node, node.attributes.filter(isMdxJsxAttribute), 200, 200);
    expect(node).toMatchSnapshot();
  });

  it("should leave node as-is if no intrinsic size is set", () => {
    const node = createElement();
    setDimension(
      node,
      node.attributes.filter(isMdxJsxAttribute),
      undefined,
      undefined
    );
    expect(node).toMatchSnapshot();
  });

  it("should do nothing with non-numeric strings", () => {
    const node = createElement("100px", "100px");
    setDimension(
      node,
      node.attributes.filter(isMdxJsxAttribute),
      undefined,
      undefined
    );
    expect(node).toMatchSnapshot();
  });
});

function createElement(
  width?: number | string,
  height?: number | string
): Hast.MdxJsxElement {
  const props: MdxJsxAttribute[] = [];

  if (width) {
    props.push({
      type: "mdxJsxAttribute",
      name: "width",
      value: `${width}`,
    });
  }

  if (height) {
    props.push({
      type: "mdxJsxAttribute",
      name: "height",
      value: `${height}`,
    });
  }
  return {
    type: "mdxJsxFlowElement",
    name: "img",
    attributes: [
      {
        type: "mdxJsxAttribute",
        name: "src",
        value: "file:123",
      },
      ...props,
    ],
    children: [],
    data: { _mdxExplicitJsx: true },
  };
}
