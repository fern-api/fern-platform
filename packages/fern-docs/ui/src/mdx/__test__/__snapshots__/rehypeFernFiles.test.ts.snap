// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setDimension > should do nothing with non-numeric strings 1`] = `
{
  "attributes": [
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
    {
      "name": "width",
      "type": "mdxJsxAttribute",
      "value": "100px",
    },
    {
      "name": "height",
      "type": "mdxJsxAttribute",
      "value": "100px",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;

exports[`setDimension > should leave node as-is if no intrinsic size is set 1`] = `
{
  "attributes": [
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;

exports[`setDimension > should preserve image with both width and height set 1`] = `
{
  "attributes": [
    {
      "name": "__assigned_imageSize",
      "type": "mdxJsxAttribute",
      "value": {
        "data": {
          "estree": {
            "body": [
              {
                "expression": {
                  "properties": [
                    {
                      "computed": false,
                      "key": {
                        "type": "Literal",
                        "value": "width",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": "100px",
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "type": "Literal",
                        "value": "height",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": "100px",
                      },
                    },
                  ],
                  "type": "ObjectExpression",
                },
                "type": "ExpressionStatement",
              },
            ],
            "sourceType": "module",
            "type": "Program",
          },
        },
        "type": "mdxJsxAttributeValueExpression",
        "value": "__expression__",
      },
    },
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
    {
      "name": "width",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
    {
      "name": "height",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;

exports[`setDimension > should properly assign height when only width is set 1`] = `
{
  "attributes": [
    {
      "name": "height",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
    {
      "name": "__assigned_imageSize",
      "type": "mdxJsxAttribute",
      "value": {
        "data": {
          "estree": {
            "body": [
              {
                "expression": {
                  "properties": [
                    {
                      "computed": false,
                      "key": {
                        "type": "Literal",
                        "value": "width",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": "100px",
                      },
                    },
                  ],
                  "type": "ObjectExpression",
                },
                "type": "ExpressionStatement",
              },
            ],
            "sourceType": "module",
            "type": "Program",
          },
        },
        "type": "mdxJsxAttributeValueExpression",
        "value": "__expression__",
      },
    },
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
    {
      "name": "width",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;

exports[`setDimension > should properly assign width when only height is set 1`] = `
{
  "attributes": [
    {
      "name": "width",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
    {
      "name": "__assigned_imageSize",
      "type": "mdxJsxAttribute",
      "value": {
        "data": {
          "estree": {
            "body": [
              {
                "expression": {
                  "properties": [
                    {
                      "computed": false,
                      "key": {
                        "type": "Literal",
                        "value": "height",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": "100px",
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "type": "Literal",
                        "value": "width",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": "auto",
                      },
                    },
                  ],
                  "type": "ObjectExpression",
                },
                "type": "ExpressionStatement",
              },
            ],
            "sourceType": "module",
            "type": "Program",
          },
        },
        "type": "mdxJsxAttributeValueExpression",
        "value": "__expression__",
      },
    },
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
    {
      "name": "height",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;

exports[`setDimension > should set width and height when neither is set 1`] = `
{
  "attributes": [
    {
      "name": "width",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
    {
      "name": "height",
      "type": "mdxJsxAttribute",
      "value": "200",
    },
    {
      "name": "src",
      "type": "mdxJsxAttribute",
      "value": "file:123",
    },
  ],
  "children": [],
  "data": {
    "_mdxExplicitJsx": true,
  },
  "name": "img",
  "type": "mdxJsxFlowElement",
}
`;
